internal/net/grpc/interceptor/server/trace/payload.go

```diff
@@ -0,0 +1,163 @@
+	"google.golang.org/grpc"
+)
+
+const (
+	grpcKindUnary  = "unary"
+	grpcKindStream = "stream"
+
+	traceAttrGRPCKind    = "grpc.kind"
+	traceAttrGRPCService = "grpc.service"
+	traceAttrGRPCMethod  = "grpc.method"
+
+	traceAttrGRPCRequestPayload  = "grpc.request.payload"
+	traceAttrGRPCResponsePayload = "grpc.response.payload"
+)
+
+func TracePayloadInterceptor() grpc.UnaryServerInterceptor {
+	return func(
+		ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (resp interface{}, err error) {
+		span := trace.FromContext(ctx)
+		if span == nil {
+			return handler(ctx, req)
+		}
+
+		service, method := parseMethod(info.FullMethod)
+		span.AddAttributes(
+			trace.StringAttribute(traceAttrGRPCKind, grpcKindUnary),
+			trace.StringAttribute(traceAttrGRPCService, service),
+			trace.StringAttribute(traceAttrGRPCMethod, method),
+		)
+
+		if reqj := marshalJSON(req); reqj != "" {
+			span.AddAttributes(
+				trace.StringAttribute(traceAttrGRPCRequestPayload, reqj),
+			)
+		}
+
+		resp, err = handler(ctx, req)
+
+		if resj := marshalJSON(resp); resj != "" {
+			span.AddAttributes(
+				trace.StringAttribute(traceAttrGRPCResponsePayload, resj),
+			)
+		}
+
+		return resp, err
+	}
+}
+
+func TracePayloadStreamInterceptor() grpc.StreamServerInterceptor {
+	return func(
+		srv interface{},
+		ss grpc.ServerStream,
+		info *grpc.StreamServerInfo,
+		handler grpc.StreamHandler,
+	) error {
+		span := trace.FromContext(ss.Context())
+		if span == nil {
+			return handler(srv, ss)
+		}
+
+		service, method := parseMethod(info.FullMethod)
+		span.AddAttributes(
+			trace.StringAttribute(traceAttrGRPCKind, grpcKindStream),
+			trace.StringAttribute(traceAttrGRPCService, service),
+			trace.StringAttribute(traceAttrGRPCMethod, method),
+		)
+
+		tss := &tracingServerStream{
+			ServerStream: ss,
+		}
+
+		err := handler(srv, tss)
+
+		span.AddAttributes(
+			trace.StringAttribute(traceAttrGRPCRequestPayload, tss.request),
+			trace.StringAttribute(traceAttrGRPCResponsePayload, tss.response),
+		)
+
+		return err
+	}
+}
+
+type tracingServerStream struct {
+	grpc.ServerStream
+	request  string
+	response string
+}
+
+func (tss *tracingServerStream) RecvMsg(m interface{}) error {
+	err := tss.ServerStream.RecvMsg(m)
+	if err == nil && tss.request == "" {
+		if reqj := marshalJSON(m); reqj != "" {
+			tss.request = reqj
+		}
+	}
+
+	return err
+}
+
+func (tss *tracingServerStream) SendMsg(m interface{}) error {
+	err := tss.ServerStream.SendMsg(m)
+	if err == nil && tss.response == "" {
+		if resj := marshalJSON(m); resj != "" {
+			tss.response = resj
+		}
+	}
+
+	return err
+}
+
+func parseMethod(fullMethod string) (service, method string) {
+	service = path.Dir(fullMethod)[1:]
+	method = path.Base(fullMethod)
+
+	return service, method
+}
+
+func marshalJSON(pbMsg interface{}) string {
+	p, ok := pbMsg.(proto.Message)
+	if !ok {
+		return ""
+	}
+
+	b := &bytes.Buffer{}
+	marshaler := &jsonpb.Marshaler{}
```

Comment:
many blog post and github community sometime said jsonpb is slow, is there any reasons that you couldn't use internal/encoding/json package instead of jsonpb?

https://github.com/golang/protobuf/issues/422
https://qiita.com/kazegusuri/items/34679ed4ad296dc338fb

Reply:
```suggestion
	json.NewEncoder(b).Encode(pbMsg)
```

Reply:
thanks. i didn't noticed that.
revised in the newer commit.